### PR TITLE
JIRA: Allow config per engagement, incl big JIRA refactor

### PR DESCRIPTION
**Pull request by valentijnscholten**
To allow users to have some flexibility in their JIRA setup  / mapping to Defect Dojo, we'd decided to allow users to the configure a JIRA project on the engagement level. This is an optional config that will override the config set on the product above it.

While implementing it I came across some logic related to this that I wanted to centralize and make it more DRY. While doing it I realized I might as well put all the JIRA related stuff in one `helper` file. And clean up some of the logic, improve connection handling, error handling, exception handling, avoid swallowing exceptions.

As I had to test "all things JIRA" I also fixed a couple of things regarding the "push all issues" stuff that has been haunting us after it was merged. Some other things found along the way and fixed:

#3205
#3206
#2664 
#3224
#3240
#3261

I also added a check when configuring a JIRA project to make sure Defect Dojo can retrieve the metadata for that JIRA project. We see quit a bit of issues on github that relate to empty metadata due to JIRA permission errors throwing a stacktrace. That is now fixed and handled more gently with error messages in the UI.

So basically this is a big refactor of the JIRA stuff, so I did extensive manual testing:

	add finding with jira enabled, without jira enabled, with push_all, without push_all
	add ad hoc finding with jira enabled, without jira enabled, with push_all, without push_all
	edit finding with jira enabled, without jira enabled, with push_all, without push_all
	upload report with jira enabled, without jira enabled, with push_all, without push_all
	reimport report with jira enabled, without jira enabled, with push_all, without push_all
	api upload report with jira enabled, without jira enabled, with push_all, without push_all
	api reupload report with jira enabled, without jira enabled, with push_all, without push_all
	api add finding with jira enabled, without jira enabled, with push_all, without push_all
	api add finding with push to jira with jira enabled, without jira enabled, with push_all, without push_all
	api add finding with push_all with jira enabled, without jira enabled, with push_all, without push_all
	push epic from engagements
	create /edit epic without jira enabled
	api patch finding with push_all with jira enabled, without jira enabled, with push_all, without push_all
	check api response for findings
        add /edit jira config for product
        add / edit jira config for engagement
        check used jira config if overridden by engagement
        webhook (added automated tests case)

There is a somewhat complex migration for which it would be really helpful if people could test that. Apart from that, it's not a PR that you can just easily approve by looking at the code for 5 minutes. Sorry for that ;-)

I might extend our test suite to actually use our jira cloud instance during the tests, but there are some issues because GHA won't expose secrets to PRs on forks.

Another thing I am not sure on is APIv1. I think it still works, but people modifying jira configuration via the API might need to tweak their scripts. Because of a small risk of breaking changes I am adding that label so people get triggered while reading the release notes or when they have an issue.

EDIT 15/11/2020:
Added a set of JIRA unittests that cover most of the APIv2 usage of push_to_jira. Details in [test_jira_config_and_pushing_api.py](https://github.com/DefectDojo/django-DefectDojo/pull/3200/files#diff-d5bfdd80b8dcb707a1ba7ec2d49a3f28290d83728e22c26976db6d0c5820263a)

### CODE REVIEW

File: dojo/api.py [hunk 1]
```diff
@@ -828,10 +828,10 @@ def validation(self):
 
 
 """
-    /api/v1/jira_configurations/
+    /api/v1/JIRA_Configurations/
```
Comment (madchap) [inline dojo/api.py:831]: Please, do not mix naming conventions. I think the original you changed was way better.
This makes me think of `is_Mitigated`. Please don't do this :-)

I know it's already in plenty of places, but let's not make it worse. And I am guessing you actually did this to match the rest.

Only 1 comment, but applies in several places.

File: dojo/api.py [hunk 2]
```diff
@@ -840,10 +840,10 @@ def validation(self):
 class JIRA_ConfResource(BaseModelResource):
 
     class Meta:
-        resource_name = 'jira_configurations'
+        resource_name = 'JIRA_Instanceigurations'
```
Comment (madchap) [inline dojo/api.py:843]: wut? :-)

**Comment by valentijnscholten** [comment]
Hold your horses. This is a DRAFT PR and has WIP written on it :-) it's full of find-replace artifacts.

File: dojo/api.py (hunk 1 above)
Comment (valentijnscholten) [reply dojo/api.py:831]: it's just a comment

**Comment by madchap** [comment]
Some notes:

Created a product with 3 engagements:
- product has JIRA configuration
- 2 of the engagements have different JIRA project configured
- 1 of the engagement has no JIRA configuration and is expected to pick-up the product level configuration.

- Expectation miss: The one engagement with no override in JIRA configuration fails to pickup the product level jira config.
Log related to that
```
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:91] found jira_project () for Engagement: prod-level-config (Nov 11, 2020)
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:91] found jira_project () for Engagement: prod-level-config (Nov 11, 2020)
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:91] found jira_project () for Engagement: prod-level-config (Nov 11, 2020)
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:118] found jira_instance None for CVE-2019-5094 - E2fsprogs-1.44.1-1ubuntu1.1(dpkg)
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:402] Trying to create a new JIRA issue for finding 26...
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:91] found jira_project () for Engagement: prod-level-config (Nov 11, 2020)
celeryworker_1  | [11/Nov/2020 10:42:22] DEBUG [dojo.jira_link.helper:118] found jira_instance None for CVE-2019-5094 - E2fsprogs-1.44.1-1ubuntu1.1(dpkg)
celeryworker_1  | [11/Nov/2020 10:42:22] ERROR [celery.app.trace:257] Task dojo.jira_link.helper.add_jira_issue[fd207979-ee2f-4910-8128-8667d31317ac] raised unexpected: AttributeError("'NoneType' object has no attribute 'default_issue_type'",)
celeryworker_1  | Traceback (most recent call last):
celeryworker_1  |   File "/usr/local/lib/python3.6/site-packages/celery/app/trace.py", line 412, in trace_task
celeryworker_1  |     R = retval = fun(*args, **kwargs)
celeryworker_1  |   File "/usr/local/lib/python3.6/site-packages/celery/app/trace.py", line 704, in __protected_call__
celeryworker_1  |     return self.run(*args, **kwargs)
celeryworker_1  |   File "/app/dojo/jira_link/helper.py", line 415, in add_jira_issue
celeryworker_1  |     'name': jira_instance.default_issue_type
celeryworker_1  | AttributeError: 'NoneType' object has no attribute 'default_issue_type'
```

- Creating new engagement with JIRA project `MAD2` and clicking "import finding shows the product level JIRA:
image

It still works as expected though, just not showing the right URL.

- For the engagement that overrides the default product level JIRA configuration, I would a small visual clue somewhere.

- When specific JIRA are specified for engagements, jira issues get created as expected :+1:

**Comment by madchap** [comment]
In my short tests, I have not noticed any regression due to refactoring. But this was a small manual test too.

**Review by madchap — CHANGES_REQUESTED** [review]
Per comments.

**Comment by valentijnscholten** [comment]
How did you create the prod level jira config? from scratch or was it an existing config? did the migration run OK?

**Comment by madchap** [comment]
> How did you create the prod level jira config? from scratch or was it an existing config? did the migration run OK?

All from scratch. I always test PR with a fresh docker-compose at HEAD of `dev`.

**Comment by valentijnscholten** [comment]
- added placeholder + help_text for engagement that the jira config is inherited from the product, unless overriden in the engagement
- fixed the display bug of the jira url on import_scan view

**Comment by madchap** [comment]
For the visual clue, could we have something on the engagement list page as well when overriden? Maybe like the little jira "bug" that's also present on the product list page?

image

I tested the migration at scale, worked fine.

**Comment by valentijnscholten** [comment]
added
image
also on list of all engagements and list of active engagements.

**Review by madchap — APPROVED** [review]
LGTM. I tested that PR on a large staging instance, everything seems to be fine, including migrations.

**Comment by Maffooch** [comment]
I did a manual test with a jira config at the product level before pull these changes. Migration seems to be fine. Claiming an inherited engagement and creating new ones worked fine as well. Overall did not see anything fishy in the logs or the UI. 
Great work!

**Review by Maffooch — APPROVED** [review]

**Pull request merged**